Userland/DynamicLoader/main.cpp

```diff
@@ -44,20 +46,73 @@ static void perform_self_relocations(auxv_t* auxvp)
         exit(1);
 }
 
-static void display_help()
+static ErrorOr<int> open_executable(StringView path)
 {
-    char const message[] =
-        R"(You have invoked `Loader.so'. This is the helper program for programs that
-use shared libraries. Special directives embedded in executables tell the
-kernel to load this program.
-
-This helper program loads the shared libraries needed by the program,
-prepares the program to run, and runs it. You do not need to invoke
-this helper program directly. If you still want to run a program with the loader,
-run: /usr/lib/Loader.so [ELF_BINARY])";
-    fprintf(stderr, "%s", message);
+    int rc = open(path.characters_without_null_termination(), O_RDONLY | O_EXEC, 0666);
+    if (rc < 0)
+        return Error::from_errno(errno);
+    int checked_fd = rc;
+    ArmedScopeGuard close_on_failure([checked_fd] {
+        close(checked_fd);
+    });
+
+    struct stat executable_stat;
+    rc = fstat(checked_fd, &executable_stat);
+    if (rc < 0)
+        return Error::from_errno(errno);
+    if (!S_ISREG(executable_stat.st_mode)) {
+        if (S_ISDIR(executable_stat.st_mode))
+            return Error::from_errno(EISDIR);
+        return Error::from_errno(EINVAL);
+    }
+
+    close_on_failure.disarm();
+    return checked_fd;
+}
+
+// FIXME: Find a better way to supply the environment variables to serenity_main.
+char** g_env;
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    bool flag_secure { false };
+    Core::ArgsParser args_parser;
+    Vector<StringView> command;
+
+    args_parser.set_general_help("Run dynamically-linked ELF executables");
+    args_parser.set_stop_on_first_non_option(true);
+    args_parser.add_option(flag_secure, "Run in secure-mode", "secure", 's');
```

Comment:
But secure mode is a property of the executable we are executing, as determined by the kernel. It's a real security feature. 

The way this is happening, if you pass --secure, I don't think it will actually pass through to calls of `secure_getenv`, which check the auxvals directly.

```
char* secure_getenv(char const* name)
{
    if (getauxval(AT_SECURE))
        return nullptr;
    return getenv(name);
}
```

So this implementation of running an app on the command line in secure mode is insufficient.

Reply:
I dropped this for now. I will try to see if it's possible to implement this in the future :)